### PR TITLE
cmd/tailscale: add web subcommand

### PR DESCRIPTION
**Pull request by crawshaw**
Used as an app frontend UI on Synology.

### CODE REVIEW

File: cmd/tailscale/cli/web.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+var tmpl = template.Must(template.New("html").Parse(webHTML))
+
+var webCmd = &ffcli.Command{
+	Name:       "web",
+	ShortUsage: "web [flags]",
+	ShortHelp:  "Run a web server for controlling Tailscale",
+
+	FlagSet: (func() *flag.FlagSet {
+		webf := flag.NewFlagSet("web", flag.ExitOnError)
+		webf.StringVar(&webArgs.addr, "addr", "localhost:8088", "local address of the web server")
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:39]: let's be consistent with `tailscale status --listen`:

```
$ tailscale status -h
USAGE
  status [-active] [-web] [-json]

FLAGS
  -active false           filter output to only peers with active sessions (not applicable to web mode)
  -browser true           Open a browser in web mode
  -json false             output in JSON format (WARNING: format subject to change)
  -listen 127.0.0.1:8384  listen address; use port 0 for automatic
  -peers true             show status of peers
  -self true              show status of local machine
  -web false              run webserver with HTML showing status
```

File: cmd/tailscale/cli/web.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+`
+
+func webHandler(w http.ResponseWriter, r *http.Request) {
+	if synoTokenRedirect(w, r) {
+		return
+	}
+
+	user, err := auth()
+	if err != nil {
+		http.Error(w, err.Error(), 403)
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:128]: agree with github actions here. I can never remember 403 vs 401 without the names. (I'm personally fine with 404, 400, 500 literals, though)

File: cmd/tailscale/cli/web.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+		return
+	}
+
+	if r.Method == "POST" {
+		w.Header().Set("Content-Type", "application/json")
+		url, err := tailscaleUp(r.Context())
+		if err != nil {
+			fmt.Fprintf(w, `{"error":%q}`, err.Error())
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:136]: I really dislike Fprintf for JSON, especially as %q doesn't exactly match what JSON requires.

```go
json.NewEncoder(w).Encode(jres{Error: err})
```

File: cmd/tailscale/cli/web.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+		fmt.Fprintf(w, `{"url":%q}`, url)
+		return
+	}
+
+	st, err := tailscale.Status(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+	}
+
+	var data struct {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:148]: make this a named type and declare it up by the `var tmpl`?

File: cmd/tailscale/cli/web.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		SynologyUser string
+		Status string
+		DeviceName string
+		IP string
+	}
+	data.SynologyUser = user
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:154]: then these could all be composite literal fields

File: cmd/tailscale/cli/web.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+	buf := new(bytes.Buffer)
+	if err := tmpl.Execute(buf, data); err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	w.Write(buf.Bytes())
+}
+
+// TODO(crawshaw): some of this is very similar to the code in 'tailscale up', can we share anything?
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:170]: I'll unify them later when I push more of this down into the localapi.

**Review by bradfitz — APPROVED** [review]

File: cmd/tailscale/cli/web.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	user, err := auth()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusForbidden)
+		return
+	}
+
+	if r.Method == "POST" {
+		w.Header().Set("Content-Type", "application/json")
+		url, err := tailscaleUp(r.Context())
+		if err != nil {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:143]: At least a local type here?

```go
type mi map[string]interface{}
```

Then below:

```go
    json.NewEncoder(w).Encode(mi{"error": err})
```

etc

File: cmd/tailscale/cli/web.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+	}
+
+	data := tmplData{
+		SynologyUser: user,
+		Status:       st.BackendState,
+		DeviceName:   st.Self.DNSName,
+	}
+	for _, ip := range st.TailscaleIPs {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:161]: 1 line fewer is:

```go
    if len(st.TailscaleIPs) != 0 {
          data.IP = st.TailscaleIPs[0].String()
    }
```

**Review by bradfitz — APPROVED** [review]

File: cmd/tailscale/cli/web.go (hunk 7 above)
Comment (crawshaw) [reply cmd/tailscale/cli/web.go:143]: Is that easier to read?

Comment (crawshaw) [reply cmd/tailscale/cli/web.go:143]: I find it more mysterious, but done.

**Pull request merged**